Region/RegionStats.cc

```diff
@@ -127,51 +143,78 @@ void RegionStats::calcHistogram(int channel, int stokes, int nBins, float minVal
-        // stats for entire region, not per channel
-        bool haveStats(getStatsValues(results, m_regionStats, subLattice, false));
-        // update message whether have stats or not
-        for (size_t i=0; i<m_regionStats.size(); ++i) {
-            // add StatisticsValue to message
-            auto statsValue = statsData.add_statistics();
-            auto statType = static_cast<CARTA::StatsType>(m_regionStats[i]);
-            statsValue->set_stats_type(statType);
-            if (!haveStats || results[i].empty()) { // region outside image or NaNs
-                if (statType==CARTA::NumPixels) {
-                    statsValue->set_value(0.0);
+        if (m_statsValid && (m_statsData.count(stokes)) && (m_statsData.at(stokes).count(channel))) {
+            // used stored stats
+            try {
+                std::vector<double> storedStats(m_statsData.at(stokes).at(channel));
+                for (size_t i=0; i<m_statsReqs.size(); ++i) {
+                    // add StatisticsValue to message
+                    auto statsValue = statsData.add_statistics();
+                    auto cartaStatType = static_cast<CARTA::StatsType>(m_statsReqs[i]);
+                    statsValue->set_stats_type(cartaStatType);
+                    statsValue->set_value(storedStats[cartaStatType-13]);
+                }
+            } catch (std::out_of_range& rangeError) {
+                // stats cleared
+                auto statsValue = statsData.add_statistics();
+                statsValue->set_stats_type(CARTA::StatsType::None);
+            }
+        } else {
+            // calculate stats
+            if (!m_statsValid)
+                m_statsData.clear();
+            // per channel = false, stats for entire region
+            bool haveStats(calcStatsValues(results, m_statsReqs, subLattice, false));
+            // update message whether have stats or not
+            for (size_t i=0; i<m_statsReqs.size(); ++i) {
+                // add StatisticsValue to message
+                auto statsValue = statsData.add_statistics();
+                auto cartaStatType = static_cast<CARTA::StatsType>(m_statsReqs[i]);
+                statsValue->set_stats_type(cartaStatType);
+                double value(0.0);
+                if (!haveStats || results[i].empty()) { // region outside image or NaNs
+                    if (cartaStatType != CARTA::NumPixels) {
+                        value = std::numeric_limits<double>::quiet_NaN();
+                    }
                 } else {
-                    statsValue->set_value(std::numeric_limits<double>::quiet_NaN());
+                    value = results[i][0];
+                }
+                statsValue->set_value(value);
+
+                // cache stats values
+                if (m_statsData[stokes][channel].empty()) { // resize vector, set to NaN
+                    m_statsData[stokes][channel].resize(CARTA::StatsType_MAX-13, // 13 is the first stat
```

Comment:
See other comment about the magic number.

Reply:
Again, carta-protobuf has StatsType_MIN = None (0), StatsType_MAX = PluginDefined (31), and StatsType_ARRAYSIZE = StatsType_MAX + 1, which are totally useless unless we have an always-half-empty vector of size 32.